File: packages/python/goldenanalysis/goldenanalysis/_regressions.py
"""Pure regression decision logic — baseline strategy + direction-aware policy.

Backend-free: operates on a list of ``(run_id, value)`` history points + the
current value. ``ReportHistory`` wires storage around this.
"""

from __future__ import annotations

import statistics
from collections.abc import Sequence

from goldenanalysis.models import Direction, RegressionPolicy


def baseline_value(history: Sequence[float], strategy: str, *, window: int = 7) -> float | None:
    """The baseline to compare the current value against.

    - ``"previous"`` / ``"last_known_good"``: the most recent historical value
      (v1: ``last_known_good`` aliases ``previous`` until a health signal exists).
    - ``"rolling_median"``: median of the last ``window`` historical values — immune
      to one noisy night, where ``previous`` would alternately flag and un-flag.
    - any other string is treated as a pinned ``run_id`` and is resolved by the
      caller (which has the run->value map); here it falls through to ``previous``.

    Returns None when there's no history to compare against.
    """
    if not history:
        return None
    if strategy == "rolling_median":
        tail = list(history[-window:])
        return float(statistics.median(tail))
    # "previous", "last_known_good", or a pinned id resolved upstream.
    return float(history[-1])


def delta_pct(baseline: float, current: float) -> float:
    if baseline == 0:
        return 0.0
    return (current - baseline) / baseline * 100.0


def is_regression(direction: Direction, baseline: float, current: float, threshold_pct: float) -> bool:
    """Direction-aware: a higher_better metric flags only on a DROP beyond the
    threshold; lower_better only on a RISE; neutral on either direction."""
    d = delta_pct(baseline, current)
    if direction == "higher_better":
        return d <= -threshold_pct
    if direction == "lower_better":
        return d >= threshold_pct
    return abs(d) >= threshold_pct


def evaluate_metric(
    *,
    key: str,
    direction: Direction,
    history: Sequence[float],
    current: float,
    strategy: str,
    window: int,
    policy: RegressionPolicy,
):
    """Return a ``Regression`` for one metric, or None when there's no baseline.

    ``flagged`` reflects the direction-aware per-metric gate; the record is always
    returned (when a baseline exists) so callers can show near-misses if they want.
    """
    from goldenanalysis.models import Regression

    base = baseline_value(history, strategy, window=window)
    if base is None:
        return None
    threshold = policy.threshold_for(key)
    return Regression(
        metric=key,
        baseline=base,
        current=current,
        delta_pct=delta_pct(base, current),
        flagged=is_regression(direction, base, current, threshold),
        direction=direction,
    )

File: packages/typescript/goldenanalysis/src/core/regressions.ts
/**
 * Cross-run regression decision logic + models (edge-safe; no node: imports).
 *
 * Parity with the Python reference
 * `packages/python/goldenanalysis/goldenanalysis/_regressions.py`. The two decisions
 * the spec's worked scenario forced: the baseline is a *strategy* (not just
 * "previous"), and thresholds are *per-metric* and respect each `Metric.direction`.
 *
 * These models are camelCase (TS-idiomatic) — they are NOT cross-language wire types
 * (the wire type is `AnalysisReport`, kept snake_case in `types.ts`).
 */

import type { Direction } from "./types.js";

/** "previous" / "rolling_median" / "last_known_good", or a pinned run_id. */
export type Baseline = "previous" | "rolling_median" | "last_known_good" | string;

/** Per-metric regression thresholds (percent); falls back to `defaultPct`. */
export interface RegressionPolicy {
  readonly defaultPct: number;
  readonly perMetric: Record<string, number>;
}

export function policyThreshold(policy: RegressionPolicy, key: string): number {
  return policy.perMetric[key] ?? policy.defaultPct;
}

export function defaultPolicy(): RegressionPolicy {
  return { defaultPct: 10, perMetric: {} };
}

export interface Regression {
  readonly metric: string;
  readonly baseline: number;
  readonly current: number;
  readonly deltaPct: number;
  readonly flagged: boolean;
  readonly direction: Direction;
}

export interface TrendSeries {
  readonly metricKey: string;
  readonly dataset: string;
  readonly points: ReadonlyArray<readonly [string, number]>; // (runId, value), oldest -> newest
}

function median(values: readonly number[]): number {
  const sorted = values.slice().sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0 ? (sorted[mid - 1]! + sorted[mid]!) / 2 : sorted[mid]!;
}

/**
 * The baseline value to compare the current value against. `rolling_median` (median
 * of the last `window`) is immune to one noisy night; `previous` / `last_known_good`
 * (v1 alias) return the most recent historical value. Empty history => null.
 */
export function baselineValue(history: readonly number[], strategy: Baseline, window = 7): number | null {
  if (history.length === 0) return null;
  if (strategy === "rolling_median") return median(history.slice(-window));
  return history[history.length - 1]!;
}

export function deltaPct(baseline: number, current: number): number {
  if (baseline === 0) return 0;
  return ((current - baseline) / baseline) * 100;
}

/** Direction-aware flag: higher_better flags on a drop; lower_better on a rise; neutral either way. */
export function isRegression(direction: Direction, baseline: number, current: number, thresholdPct: number): boolean {
  const d = deltaPct(baseline, current);
  if (direction === "higher_better") return d <= -thresholdPct;
  if (direction === "lower_better") return d >= thresholdPct;
  return Math.abs(d) >= thresholdPct;
}
